go/kbfs/libkbfs/subscription_manager.go

```diff
@@ -0,0 +1,366 @@
+// Copyright 2019 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"errors"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/kbfs/data"
+	"github.com/keybase/client/go/kbfs/tlfhandle"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"golang.org/x/net/context"
+	"golang.org/x/time/rate"
+)
+
+// userPath is always the full path including the /keybase prefix, but may
+// not be canonical or cleaned. The goal is to track whatever the user of this
+// type is dealing with without needing them to know if a path is canonicalized
+// at any time.
+// Examples:
+//   "/keybase/public/karlthefog@twitter/dir
+//   "/keybase/team/keybase/dir/../file"
+type userPath string
+
+// cleanInTlfPath is clean path rooted at a TLF, and it's what we get
+// from Node.GetPathPlaintextSansTlf().
+// Examples, considering TLF /keybase/private/user1,user2:
+//   "/foo/bar" (representing /keybsae/private/user1,user2/foo/bar)
+//   "/"        (representing /keybsae/private/user1,user2)
+type cleanInTlfPath string
+
+func getCleanInTlfPath(p *parsedPath) cleanInTlfPath {
+	return cleanInTlfPath(path.Clean(p.rawInTlfPath))
+}
+
+type debouncedNotify struct {
+	notify   func()
+	shutdown func()
+}
+
+func getChSender(ch chan<- struct{}, blocking bool) func() {
+	if blocking {
```

Comment:
I don't see how these "blocking" changes address my issue, which was with the following scenario when a non-Inf rate limit is set:

1. A node is changed, sending on `ch`.
2. The main `for` loop in the `debounce()` goroutine is triggered, `do()` is called, and then `limiter.Wait()` is entered, which causes it to wait for the interval.  (Maybe blocking in `Wait()` requires one more change; if so just pretend I wrote that out too.)
3. The GUI gets the notification, calls into KBFS and gets the new state of the node.
4. The node is changed again, and tries to send another notification, which is dropped because `ch` isn't trying to receive at the moment.
5. The node is never changed again.

Now the GUI will be stuck displaying the old state, because it never got the last notification.  That's why I think `ch` needs to have a buffer length of 1, so if a notification comes in during the wait period, it will eventually get sent.  You absolutely don't want to drop that notification, you just want to rate limit it so that not too many come at once, right?